lib/hanami/action/contract.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+module Hanami
+  class Action
+    # A wrapper for defining validation rules using Dry::Validation. This class essentially
+    # wraps a Dry::Validation::Contract and acts as a proxy to actually use Dry gem
+    #
+    # Defined via the `contract` block in an action class.
+    # Although more complex domain-specific validations, or validations concerned with things such as uniqueness
+    # are usually better performed at layers deeper than your HTTP actions, Contract still provides helpful features
+    # that you can use without contravening the advice form above.
+    #
+    # @since 2.2.0
+    class Contract
+      # A wrapper for the result of a contract validation
+      # @since 2.2.0
+      # @api private
+      class Result < SimpleDelegator
+        # @since 2.0.0
+        # @api private
+        def output
+          __getobj__.to_h
+        end
+
+        # @since 2.0.0
+        # @api private
+        def messages
+          __getobj__.errors.to_h
+        end
+      end
+
+      # @attr_reader env [Hash] the Rack env
+      #
+      # @since 2.2.0
+      # @api private
+      attr_reader :env
+
+      # Define a contract for the given action
+      #
+      # @param blk [Proc] the block to define the contract, including [Params] as a contract schema and connected rules
+      #
+      # @since 2.2.0
+      # @api private
+      #
+      # @example
+      # class Create < Hanami::Action
+      #   contract do
+      #     params do
+      #       required(:birth_date).value(:date)
+      #     end
+      #     rule(:birth_date) do
+      #       key.failure('you must be 18 years or older to register') if value > Date.today - 18.years
+      #     end
+      #   end
+      #
+      #   def handle(req, *)
+      #     halt 400 unless req.contract.call.errors.empty?
+      #     # ...
+      #   end
+      # end
+      def self.contract(&blk)
+        validations(&blk || -> {})
```

Comment:
Can't we just do `@_validator = Dry::Validation::Contract.build(&blk)` here and remove `validations` method?

Reply:
You're right, it was pointless.

Thanks for the review (and while I am at it - for your OS work, that I've used extensively over the years and greatly admired). Applied all changes now.

Reply:
Thanks @krzykamil 😃